hazelcast/src/main/java/com/hazelcast/sql/impl/client/QueryClientStateRegistry.java

```diff
@@ -0,0 +1,167 @@
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.client;
+
+import com.hazelcast.internal.serialization.Data;
+import com.hazelcast.internal.serialization.InternalSerializationService;
+import com.hazelcast.sql.impl.QueryException;
+import com.hazelcast.sql.SqlRow;
+import com.hazelcast.sql.impl.QueryId;
+import com.hazelcast.sql.impl.SqlResultImpl;
+import com.hazelcast.sql.impl.SqlRowImpl;
+import com.hazelcast.sql.impl.row.Row;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Registry of active client cursors.
+ */
+public class QueryClientStateRegistry {
+    /** Registered client cursors. */
+    private final ConcurrentHashMap<QueryId, QueryClientState> clientCursors = new ConcurrentHashMap<>();
+
+    public SqlPage registerAndFetch(
+        UUID clientId,
+        SqlResultImpl cursor,
+        int cursorBufferSize,
+        InternalSerializationService serializationService
+    ) {
+        QueryClientState clientCursor = new QueryClientState(clientId, cursor);
+
+        SqlPage page = fetchInternal(clientCursor, cursorBufferSize, serializationService);
+
+        if (!page.isLast()) {
+            clientCursors.put(cursor.getQueryId(), clientCursor);
+        }
+
+        return page;
+    }
+
+    public SqlPage fetch(
+        UUID clientId,
+        QueryId queryId,
+        int cursorBufferSize,
+        InternalSerializationService serializationService
+    ) {
+        QueryClientState clientCursor = getClientCursor(clientId, queryId);
+
+        if (clientCursor == null) {
+            throw QueryException.error("Query cursor is not found (closed?): " + queryId);
+        }
+
+        SqlPage page = fetchInternal(clientCursor, cursorBufferSize, serializationService);
+
+        if (page.isLast()) {
+            deleteClientCursor(clientCursor);
+        }
+
+        return page;
+    }
+
+    private SqlPage fetchInternal(
+        QueryClientState clientCursor,
+        int cursorBufferSize,
+        InternalSerializationService serializationService
+    ) {
+        Iterator<SqlRow> iterator = clientCursor.getIterator();
+
+        List<Data> page = new ArrayList<>(cursorBufferSize);
+        boolean last = fetchPage(iterator, page, cursorBufferSize, serializationService);
+
+        if (last) {
+            deleteClientCursor(clientCursor);
+        }
+
+        return new SqlPage(page, last);
+    }
+
+    private boolean fetchPage(
+        Iterator<SqlRow> iterator,
+        List<Data> page,
+        int cursorBufferSize,
+        InternalSerializationService serializationService
+    ) {
+        while (iterator.hasNext()) {
+            SqlRow row = iterator.next();
+            Row rowInternal = ((SqlRowImpl) row).getDelegate();
+            Data rowData = serializationService.toData(rowInternal);
```

Comment:
This does not seem correct. We are not using Data for builtin `IdentifiedDataSerializable` objects on our protocol as much as possible. It is hard to define what is Data in the protocol. More importantly, they don't support evolution, it is very likely to make a mistake. 

I see that `Row` is an interface. Which `Row` implementations can we return to the client?

Reply:
I'll change it to `List<Data>`. There is no model for a row at the moment.

Reply:
Fixed. Now `List<Data>` is used to represent row values. Please take a look.